contracts/integrations/nevermined/AgentMechSubscription.sol

```diff
@@ -0,0 +1,143 @@
+
+    // Subscription NFT
+    address public subscriptionNFT;
+    // Subscription token Id
+    uint256 public subscriptionTokenId;
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    /// @dev AgentMechSubscription constructor.
+    /// @param _token Address of the token registry contract.
+    /// @param _tokenId The token ID.
+    /// @param _creditsPerRequest Number of credits to pay for request via subscription.
+    /// @param _subscriptionNFT Subscription address.
+    /// @param _subscriptionTokenId Subscription token Id.
+    constructor(
+        address _token,
+        uint256 _tokenId,
+        uint256 _creditsPerRequest,
+        address _subscriptionNFT,
+        uint256 _subscriptionTokenId
+    )
+        AgentMech(_token, _tokenId, _creditsPerRequest)
+    {
+        // Check for the subscription address
+        if (_subscriptionNFT == address(0)) {
+            revert ZeroSubscriptionAddress();
+        }
+
+        // Check for the subscription token Id
+        if (_subscriptionTokenId == 0) {
+            revert ZeroTokenId();
+        }
+
+        subscriptionNFT = _subscriptionNFT;
+        subscriptionTokenId = _subscriptionTokenId;
+    }
+
+    /// @dev Performs actions before the request is posted.
+    /// @param amount Amount of payment in wei.
+    function _preRequest(uint256 amount, uint256, bytes memory) internal override {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Check that there is no incoming deposit
+        if (amount > 0) {
+            revert NoDepositAllowed(amount);
+        }
+
+        // Check for the number of credits available in the subscription vs total number of credits needed
+        uint256 creditsBalance = IERC1155(subscriptionNFT).balanceOf(msg.sender, subscriptionTokenId);
+        uint256 numUndeliveredRequests = mapUndeliveredRequestsCounts[msg.sender];
+        uint256 creditsPerPendingRequests = (numUndeliveredRequests + 1) * price;
+        if (creditsBalance < creditsPerPendingRequests) {
+            revert NotEnoughCredits(creditsBalance, creditsPerPendingRequests);
+        }
+
+        _locked = 1;
+    }
+
+    /// @dev Performs actions before the delivery of a request.
+    /// @param requestId Request Id.
+    /// @param data Self-descriptive opaque data-blob.
+    /// @return requestData Data for the request processing.
+    function _preDeliver(uint256 requestId, bytes memory data) internal override returns (bytes memory requestData) {
+        // Reentrancy guard
+        if (_locked > 1) {
+            revert ReentrancyGuard();
+        }
+        _locked = 2;
+
+        // Burn credits of the request Id sender upon delivery
+        IERC1155(subscriptionNFT).burn(mapRequestAddresses[requestId], subscriptionTokenId, price);
+
+        // Return the request data
+        requestData = data;
+
+        _locked = 1;
+    }
+
```

Comment:
What I'm confused about here is that now the price is still static. The idea was that requests can be dynamically priced...